File: lib/utils/csvGenerator.ts
/**
 * CSV 생성 유틸리티
 * 
 * 간/비장 분석 결과를 CSV 형식으로 생성합니다.
 * 백엔드 API 대신 클라이언트 사이드에서 직접 CSV를 생성할 때 사용합니다.
 */

import { CSV_COLUMNS, CSVExportData } from '@/lib/ctTypes';

// 백엔드 API URL (환경변수로 설정 가능)
export const API_BASE_URL = process.env.NEXT_PUBLIC_API_URL || 'http://localhost:8000';

/**
 * 값을 CSV 형식으로 포맷합니다.
 */
function formatValue(value: number | string | null | undefined): string {
  if (value === null || value === undefined) {
    return '';
  }
  if (typeof value === 'number') {
    return value.toFixed(4);
  }
  return String(value);
}

/**
 * CSV 행 데이터를 생성합니다.
 */
function createRow(
  patientId: string,
  studyId: string | undefined,
  organ: 'liver' | 'spleen',
  data: {
    volume_ml?: number | null;
    mean_HU?: number | null;
    std_HU?: number | null;
    min_HU?: number | null;
    max_HU?: number | null;
    p10_HU?: number | null;
    p90_HU?: number | null;
    GLCM_contrast?: number | null;
    GLCM_homogeneity?: number | null;
    GLRLM_LRE?: number | null;
    GLSZM_ZE?: number | null;
  }
): string[] {
  return [
    patientId,
    studyId || '',
    organ,
    formatValue(data.volume_ml),
    formatValue(data.mean_HU),
    formatValue(data.std_HU),
    formatValue(data.min_HU),
    formatValue(data.max_HU),
    formatValue(data.p10_HU),
    formatValue(data.p90_HU),
    formatValue(data.GLCM_contrast),
    formatValue(data.GLCM_homogeneity),
    formatValue(data.GLRLM_LRE),
    formatValue(data.GLSZM_ZE),
  ];
}

/**
 * CSV 문자열을 생성합니다.
 */
export function generateCSVContent(data: CSVExportData): string {
  const rows: string[][] = [];
  
  // 헤더 추가
  rows.push([...CSV_COLUMNS]);
  
  // 간 데이터 추가
  if (data.liver_volume_ml !== null && data.liver_volume_ml !== undefined) {
    rows.push(createRow(data.patient_id, data.study_id, 'liver', {
      volume_ml: data.liver_volume_ml,
      mean_HU: data.liver_mean_hu,
      std_HU: data.liver_std_hu,
      min_HU: data.liver_min_hu,
      max_HU: data.liver_max_hu,
      p10_HU: data.liver_p10_hu,
      p90_HU: data.liver_p90_hu,
      GLCM_contrast: data.liver_glcm_contrast,
      GLCM_homogeneity: data.liver_glcm_homogeneity,
      GLRLM_LRE: data.liver_glrlm_lre,
      GLSZM_ZE: data.liver_glszm_ze,
    }));
  }
  
  // 비장 데이터 추가
  if (data.spleen_volume_ml !== null && data.spleen_volume_ml !== undefined) {
    rows.push(createRow(data.patient_id, data.study_id, 'spleen', {
      volume_ml: data.spleen_volume_ml,
      mean_HU: data.spleen_mean_hu,
      std_HU: data.spleen_std_hu,
      min_HU: data.spleen_min_hu,
      max_HU: data.spleen_max_hu,
      p10_HU: data.spleen_p10_hu,
      p90_HU: data.spleen_p90_hu,
      GLCM_contrast: data.spleen_glcm_contrast,
      GLCM_homogeneity: data.spleen_glcm_homogeneity,
      GLRLM_LRE: data.spleen_glrlm_lre,
      GLSZM_ZE: data.spleen_glszm_ze,
    }));
  }
  
  // CSV 문자열 생성
  return rows.map(row => 
    row.map(cell => {
      // 쉼표나 따옴표가 포함된 경우 따옴표로 감싸기
      if (cell.includes(',') || cell.includes('"') || cell.includes('\n')) {
        return `"${cell.replace(/"/g, '""')}"`;
      }
      return cell;
    }).join(',')
  ).join('\n');
}

/**
 * CSV 파일을 Blob으로 생성합니다 (UTF-8 BOM 포함).
 * 엑셀에서 한글/영문이 깨지지 않도록 BOM을 추가합니다.
 */
export function generateCSVBlob(data: CSVExportData): Blob {
  const csvContent = generateCSVContent(data);
  // UTF-8 BOM
  const bom = new Uint8Array([0xEF, 0xBB, 0xBF]);
  const csvBytes = new TextEncoder().encode(csvContent);
  
  return new Blob([bom, csvBytes], { type: 'text/csv;charset=utf-8' });
}

/**
 * CSV 파일을 다운로드합니다.
 */
export function downloadCSV(data: CSVExportData, filename?: string): void {
  const blob = generateCSVBlob(data);
  const timestamp = new Date().toISOString().slice(0, 10).replace(/-/g, '');
  const defaultFilename = `liver_spleen_analysis_${data.patient_id}_${timestamp}.csv`;
  
  const url = URL.createObjectURL(blob);
  const a = document.createElement('a');
  a.href = url;
  a.download = filename || defaultFilename;
  document.body.appendChild(a);
  a.click();
  document.body.removeChild(a);
  URL.revokeObjectURL(url);
}

/**
 * 백엔드 API를 통해 CSV를 다운로드합니다.
 * API 호출이 실패하면 클라이언트 사이드에서 직접 생성합니다.
 */
export async function downloadCSVFromAPI(data: CSVExportData): Promise<void> {
  // 쿼리 파라미터 생성
  const params = new URLSearchParams();
  params.append('patient_id', data.patient_id);
  
  if (data.study_id) params.append('study_id', data.study_id);
  
  // 간 데이터
  if (data.liver_volume_ml != null) params.append('liver_volume_ml', String(data.liver_volume_ml));
  if (data.liver_mean_hu != null) params.append('liver_mean_hu', String(data.liver_mean_hu));
  if (data.liver_std_hu != null) params.append('liver_std_hu', String(data.liver_std_hu));
  if (data.liver_min_hu != null) params.append('liver_min_hu', String(data.liver_min_hu));
  if (data.liver_max_hu != null) params.append('liver_max_hu', String(data.liver_max_hu));
  if (data.liver_p10_hu != null) params.append('liver_p10_hu', String(data.liver_p10_hu));
  if (data.liver_p90_hu != null) params.append('liver_p90_hu', String(data.liver_p90_hu));
  if (data.liver_glcm_contrast != null) params.append('liver_glcm_contrast', String(data.liver_glcm_contrast));
  if (data.liver_glcm_homogeneity != null) params.append('liver_glcm_homogeneity', String(data.liver_glcm_homogeneity));
  if (data.liver_glrlm_lre != null) params.append('liver_glrlm_lre', String(data.liver_glrlm_lre));
  if (data.liver_glszm_ze != null) params.append('liver_glszm_ze', String(data.liver_glszm_ze));
  
  // 비장 데이터
  if (data.spleen_volume_ml != null) params.append('spleen_volume_ml', String(data.spleen_volume_ml));
  if (data.spleen_mean_hu != null) params.append('spleen_mean_hu', String(data.spleen_mean_hu));
  if (data.spleen_std_hu != null) params.append('spleen_std_hu', String(data.spleen_std_hu));
  if (data.spleen_min_hu != null) params.append('spleen_min_hu', String(data.spleen_min_hu));
  if (data.spleen_max_hu != null) params.append('spleen_max_hu', String(data.spleen_max_hu));
  if (data.spleen_p10_hu != null) params.append('spleen_p10_hu', String(data.spleen_p10_hu));
  if (data.spleen_p90_hu != null) params.append('spleen_p90_hu', String(data.spleen_p90_hu));
  if (data.spleen_glcm_contrast != null) params.append('spleen_glcm_contrast', String(data.spleen_glcm_contrast));
  if (data.spleen_glcm_homogeneity != null) params.append('spleen_glcm_homogeneity', String(data.spleen_glcm_homogeneity));
  if (data.spleen_glrlm_lre != null) params.append('spleen_glrlm_lre', String(data.spleen_glrlm_lre));
  if (data.spleen_glszm_ze != null) params.append('spleen_glszm_ze', String(data.spleen_glszm_ze));

  try {
    const response = await fetch(
      `${API_BASE_URL}/api/abdomen/liver-spleen/csv?${params.toString()}`,
      {
        method: 'GET',
        headers: {
          'Accept': 'text/csv',
        },
      }
    );
    
    if (!response.ok) {
      throw new Error(`API 응답 오류: ${response.status}`);
    }
    
    // Content-Disposition 헤더에서 파일명 추출
    const contentDisposition = response.headers.get('Content-Disposition');
    let filename = `liver_spleen_analysis_${data.patient_id}.csv`;
    if (contentDisposition) {
      const match = contentDisposition.match(/filename="?([^"]+)"?/);
      if (match) {
        filename = match[1];
      }
    }
    
    // Blob으로 변환하여 다운로드
    const blob = await response.blob();
    const url = URL.createObjectURL(blob);
    const a = document.createElement('a');
    a.href = url;
    a.download = filename;
    document.body.appendChild(a);
    a.click();
    document.body.removeChild(a);
    URL.revokeObjectURL(url);
    
  } catch (error) {
    console.warn('백엔드 API 호출 실패, 클라이언트 사이드에서 CSV 생성:', error);
    // Fallback: 클라이언트 사이드에서 직접 생성
    downloadCSV(data);
  }
}

File: backend/utils/csv_generator.py
"""
CSV 파일 생성 유틸리티

간/비장 분석 결과를 CSV 형식으로 생성합니다.
"""
import io
import csv
from typing import Dict, Any, Optional, List
from ..models.schemas import CSV_COLUMNS


def generate_csv_content(
    patient_id: str,
    study_id: Optional[str],
    liver_data: Dict[str, Any],
    spleen_data: Dict[str, Any],
) -> str:
    """
    간/비장 분석 결과를 CSV 문자열로 생성합니다.
    
    Args:
        patient_id: 환자 ID
        study_id: 검사/스터디 ID
        liver_data: 간 분석 데이터
        spleen_data: 비장 분석 데이터
    
    Returns:
        CSV 형식 문자열 (UTF-8 BOM 포함)
    """
    output = io.StringIO()
    writer = csv.writer(output, quoting=csv.QUOTE_MINIMAL)
    
    # 헤더 작성
    writer.writerow(CSV_COLUMNS)
    
    # 간 데이터 행 작성
    if liver_data:
        liver_row = _create_row(patient_id, study_id, "liver", liver_data)
        writer.writerow(liver_row)
    
    # 비장 데이터 행 작성
    if spleen_data:
        spleen_row = _create_row(patient_id, study_id, "spleen", spleen_data)
        writer.writerow(spleen_row)
    
    return output.getvalue()


def _create_row(
    patient_id: str,
    study_id: Optional[str],
    organ: str,
    data: Dict[str, Any]
) -> List[Any]:
    """
    단일 장기의 CSV 행을 생성합니다.
    
    Args:
        patient_id: 환자 ID
        study_id: 검사/스터디 ID
        organ: 장기 이름 (liver/spleen)
        data: 장기 분석 데이터
    
    Returns:
        CSV 행 데이터 리스트
    """
    def format_value(value: Any) -> str:
        """값을 CSV 형식으로 포맷합니다."""
        if value is None:
            return ""
        if isinstance(value, float):
            return f"{value:.4f}"
        return str(value)
    
    return [
        patient_id,
        study_id or "",
        organ,
        format_value(data.get("volume_ml")),
        format_value(data.get("mean_HU")),
        format_value(data.get("std_HU")),
        format_value(data.get("min_HU")),
        format_value(data.get("max_HU")),
        format_value(data.get("p10_HU")),
        format_value(data.get("p90_HU")),
        format_value(data.get("GLCM_contrast")),
        format_value(data.get("GLCM_homogeneity")),
        format_value(data.get("GLRLM_LRE")),
        format_value(data.get("GLSZM_ZE")),
    ]


def generate_csv_bytes(
    patient_id: str,
    study_id: Optional[str],
    liver_data: Dict[str, Any],
    spleen_data: Dict[str, Any],
) -> bytes:
    """
    CSV 내용을 바이트로 생성합니다 (UTF-8 BOM 포함).
    
    엑셀에서 한글/영문이 깨지지 않도록 BOM을 추가합니다.
    
    Args:
        patient_id: 환자 ID
        study_id: 검사/스터디 ID
        liver_data: 간 분석 데이터
        spleen_data: 비장 분석 데이터
    
    Returns:
        UTF-8 BOM이 포함된 CSV 바이트
    """
    csv_content = generate_csv_content(patient_id, study_id, liver_data, spleen_data)
    
    # UTF-8 BOM (0xEF, 0xBB, 0xBF) 추가
    bom = b'\xef\xbb\xbf'
    return bom + csv_content.encode('utf-8')


def create_csv_from_request(
    patient_id: str,
    study_id: Optional[str] = None,
    liver_volume_ml: Optional[float] = None,
    liver_mean_hu: Optional[float] = None,
    liver_std_hu: Optional[float] = None,
    liver_min_hu: Optional[float] = None,
    liver_max_hu: Optional[float] = None,
    liver_p10_hu: Optional[float] = None,
    liver_p90_hu: Optional[float] = None,
    liver_glcm_contrast: Optional[float] = None,
    liver_glcm_homogeneity: Optional[float] = None,
    liver_glrlm_lre: Optional[float] = None,
    liver_glszm_ze: Optional[float] = None,
    spleen_volume_ml: Optional[float] = None,
    spleen_mean_hu: Optional[float] = None,
    spleen_std_hu: Optional[float] = None,
    spleen_min_hu: Optional[float] = None,
    spleen_max_hu: Optional[float] = None,
    spleen_p10_hu: Optional[float] = None,
    spleen_p90_hu: Optional[float] = None,
    spleen_glcm_contrast: Optional[float] = None,
    spleen_glcm_homogeneity: Optional[float] = None,
    spleen_glrlm_lre: Optional[float] = None,
    spleen_glszm_ze: Optional[float] = None,
) -> bytes:
    """
    개별 파라미터로부터 CSV 바이트를 생성합니다.
    
    프론트엔드에서 전달된 데이터로 CSV를 생성할 때 사용합니다.
    """
    liver_data = {}
    if liver_volume_ml is not None:
        liver_data = {
            "volume_ml": liver_volume_ml,
            "mean_HU": liver_mean_hu,
            "std_HU": liver_std_hu,
            "min_HU": liver_min_hu,
            "max_HU": liver_max_hu,
            "p10_HU": liver_p10_hu,
            "p90_HU": liver_p90_hu,
            "GLCM_contrast": liver_glcm_contrast,
            "GLCM_homogeneity": liver_glcm_homogeneity,
            "GLRLM_LRE": liver_glrlm_lre,
            "GLSZM_ZE": liver_glszm_ze,
        }
    
    spleen_data = {}
    if spleen_volume_ml is not None:
        spleen_data = {
            "volume_ml": spleen_volume_ml,
            "mean_HU": spleen_mean_hu,
            "std_HU": spleen_std_hu,
            "min_HU": spleen_min_hu,
            "max_HU": spleen_max_hu,
            "p10_HU": spleen_p10_hu,
            "p90_HU": spleen_p90_hu,
            "GLCM_contrast": spleen_glcm_contrast,
            "GLCM_homogeneity": spleen_glcm_homogeneity,
            "GLRLM_LRE": spleen_glrlm_lre,
            "GLSZM_ZE": spleen_glszm_ze,
        }
    
    return generate_csv_bytes(patient_id, study_id, liver_data, spleen_data)

